File: sdk/react-native/android/src/main/java/com/gomailer/gomailer/GoMailerModule.kt
package com.gomailer.gomailer

import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.net.HttpURLConnection
import java.net.URL
import java.io.OutputStreamWriter
import java.text.SimpleDateFormat
import java.util.*
import org.json.JSONObject
import com.google.firebase.messaging.FirebaseMessaging

class GoMailerModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private var apiKey: String? = null
    private var baseUrl: String = "https://api.go-mailer.com/v1"
    private var currentEmail: String? = null
    private var currentDeviceToken: String? = null

    override fun getName(): String {
        return "GoMailerModule"
    }

    @ReactMethod
    fun initialize(config: ReadableMap, promise: Promise) {
        try {
            apiKey = config.getString("apiKey")
            if (config.hasKey("baseUrl")) {
                baseUrl = config.getString("baseUrl") ?: baseUrl
            }
            
            Log.d("GoMailer", "🚀 Initializing Go Mailer module")
            Log.d("GoMailer", "✅ API Key set: ${apiKey?.take(10)}...")
            Log.d("GoMailer", "🌐 Base URL set: $baseUrl")
            
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to initialize: ${e.message}")
            promise.reject("INIT_ERROR", "Failed to initialize Go Mailer", e)
        }
    }

    @ReactMethod
    fun setUser(user: ReadableMap, promise: Promise) {
        try {
            currentEmail = user.getString("email")
            Log.d("GoMailer", "👤 Setting user data")
            Log.d("GoMailer", "📧 User email set: $currentEmail")
            
            // Send user data to backend
            sendUserToServer(currentEmail ?: "")
            
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to set user: ${e.message}")
            promise.reject("SET_USER_ERROR", "Failed to set user", e)
        }
    }

    @ReactMethod
    fun registerForPushNotifications(params: ReadableMap, promise: Promise) {
        try {
            val email = params.getString("email") ?: currentEmail
            currentEmail = email
            
            Log.d("GoMailer", "📱 Registering for push notifications")
            Log.d("GoMailer", "📧 Email for registration: $email")
            
            // Get FCM token
            FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
                if (!task.isSuccessful) {
                    Log.w("GoMailer", "Fetching FCM registration token failed", task.exception)
                    return@addOnCompleteListener
                }

                // Get new FCM registration token
                val token = task.result
                currentDeviceToken = token
                
                Log.d("GoMailer", "✅ FCM token: ${token?.take(20)}...")
                
                // Send device token to backend
                sendDeviceTokenToServer(token ?: "", email ?: "")
            }
            
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to register for push notifications: ${e.message}")
            promise.reject("REGISTER_ERROR", "Failed to register for push notifications", e)
        }
    }

    @ReactMethod
    fun getDeviceToken(promise: Promise) {
        try {
            Log.d("GoMailer", "🔑 Getting device token")
            
            if (currentDeviceToken != null) {
                Log.d("GoMailer", "✅ Returning stored device token: ${currentDeviceToken?.take(20)}...")
                val result = Arguments.createMap().apply {
                    putString("token", currentDeviceToken)
                }
                promise.resolve(result)
            } else {
                Log.d("GoMailer", "⚠️ No device token available yet")
                promise.resolve(null)
            }
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to get device token: ${e.message}")
            promise.reject("GET_TOKEN_ERROR", "Failed to get device token", e)
        }
    }

    @ReactMethod
    fun trackEvent(params: ReadableMap, promise: Promise) {
        try {
            val eventName = params.getString("eventName") ?: "unknown"
            val properties = params.getMap("properties")
            
            Log.d("GoMailer", "📊 Tracking event: $eventName")
            Log.d("GoMailer", "📊 Properties: $properties")
            
            // Send event to backend
            sendEventToServer(eventName, properties)
            
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to track event: ${e.message}")
            promise.reject("TRACK_EVENT_ERROR", "Failed to track event", e)
        }
    }

    @ReactMethod
    fun trackNotificationClick(params: ReadableMap, promise: Promise) {
        try {
            val notificationId = params.getString("notificationId") ?: "unknown"
            val title = params.getString("title") ?: "Unknown"
            val body = params.getString("body") ?: ""
            val email = params.getString("email") ?: currentEmail ?: ""
            
            Log.d("GoMailer", "👆 Tracking notification click: $notificationId")
            
            val properties = Arguments.createMap().apply {
                putString("notification_id", notificationId)
                putString("notification_title", title)
                putString("notification_body", body)
                putString("clicked_timestamp", SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).format(Date()))
                putString("platform", "android")
            }
            
            sendEventToServer("notification_clicked", properties)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to track notification click: ${e.message}")
            promise.reject("TRACK_NOTIFICATION_CLICK_ERROR", "Failed to track notification click", e)
        }
    }

    @ReactMethod
    fun setUserAttributes(attributes: ReadableMap, promise: Promise) {
        try {
            Log.d("GoMailer", "🏷️ Setting user attributes")
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to set user attributes: ${e.message}")
            promise.reject("SET_ATTRIBUTES_ERROR", "Failed to set user attributes", e)
        }
    }

    @ReactMethod
    fun addUserTags(tags: ReadableArray, promise: Promise) {
        try {
            Log.d("GoMailer", "➕ Adding user tags")
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to add user tags: ${e.message}")
            promise.reject("ADD_TAGS_ERROR", "Failed to add user tags", e)
        }
    }

    @ReactMethod
    fun removeUserTags(tags: ReadableArray, promise: Promise) {
        try {
            Log.d("GoMailer", "➖ Removing user tags")
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to remove user tags: ${e.message}")
            promise.reject("REMOVE_TAGS_ERROR", "Failed to remove user tags", e)
        }
    }

    @ReactMethod
    fun setAnalyticsEnabled(enabled: Boolean, promise: Promise) {
        try {
            Log.d("GoMailer", "📈 Setting analytics enabled: $enabled")
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to set analytics enabled: ${e.message}")
            promise.reject("SET_ANALYTICS_ERROR", "Failed to set analytics enabled", e)
        }
    }

    @ReactMethod
    fun setLogLevel(level: String, promise: Promise) {
        try {
            Log.d("GoMailer", "📝 Setting log level: $level")
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("GoMailer", "❌ Failed to set log level: ${e.message}")
            promise.reject("SET_LOG_LEVEL_ERROR", "Failed to set log level", e)
        }
    }

    private fun sendUserToServer(email: String) {
        Thread {
            try {
                val url = URL("$baseUrl/contacts")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                apiKey?.let { connection.setRequestProperty("Authorization", "Bearer $it") }
                connection.doOutput = true

                val body = JSONObject().apply {
                    put("email", email)
                    put("gm_mobi_push", JSONObject().apply {
                        put("platform", "android")
                        put("bundleId", "com.push_test_react_native_example")
                        put("appVersion", "1.0.0")
                        put("timestamp", java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", java.util.Locale.US).format(java.util.Date()))
                    })
                }

                Log.d("GoMailer", "📤 Sending user data to: $url")
                Log.d("GoMailer", "📤 Request body: $body")

                val outputStream = connection.outputStream
                val writer = OutputStreamWriter(outputStream)
                writer.write(body.toString())
                writer.flush()
                writer.close()

                val responseCode = connection.responseCode
                Log.d("GoMailer", "📤 User data sent to backend. Status: $responseCode")

                if (responseCode >= 400) {
                    val errorStream = connection.errorStream
                    val errorResponse = errorStream?.bufferedReader()?.use { it.readText() }
                    Log.e("GoMailer", "❌ Server error response: $errorResponse")
                }

                connection.disconnect()
            } catch (e: Exception) {
                Log.e("GoMailer", "❌ Failed to send user data: ${e.message}")
            }
        }.start()
    }

    private fun sendDeviceTokenToServer(token: String, email: String) {
        Thread {
            try {
                val url = URL("$baseUrl/contacts")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                apiKey?.let { connection.setRequestProperty("Authorization", "Bearer $it") }
                connection.doOutput = true

                val body = JSONObject().apply {
                    put("email", email)
                    put("gm_mobi_push", JSONObject().apply {
                        put("deviceToken", token)
                        put("platform", "android")
                        put("bundleId", "com.push_test_react_native_example")
                        put("appVersion", "1.0.0")
                        put("timestamp", java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", java.util.Locale.US).format(java.util.Date()))
                    })
                }

                Log.d("GoMailer", "📤 Sending device token to: $url")
                Log.d("GoMailer", "📤 Request body: $body")

                val outputStream = connection.outputStream
                val writer = OutputStreamWriter(outputStream)
                writer.write(body.toString())
                writer.flush()
                writer.close()

                val responseCode = connection.responseCode
                Log.d("GoMailer", "📤 Device token sent to backend. Status: $responseCode")

                if (responseCode >= 400) {
                    val errorStream = connection.errorStream
                    val errorResponse = errorStream?.bufferedReader()?.use { it.readText() }
                    Log.e("GoMailer", "❌ Server error response: $errorResponse")
                }

                connection.disconnect()
            } catch (e: Exception) {
                Log.e("GoMailer", "❌ Failed to send device token: ${e.message}")
            }
        }.start()
    }

    private fun sendEvent(eventName: String, data: WritableMap) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, data)
    }

    private fun sendEventToServer(eventName: String, properties: ReadableMap?) {
        Thread {
            try {
                val url = URL("$baseUrl/events/push")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.setRequestProperty("Authorization", "Bearer $apiKey")
                connection.doOutput = true

                val body = JSONObject().apply {
                    put("email", currentEmail ?: "")
                    put("eventName", eventName)
                    put("properties", properties?.toString() ?: "{}")
                    put("platform", "android")
                    put("appVersion", "1.0.0")
                    put("timestamp", java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").format(java.util.Date()))
                }

                val outputStream = connection.outputStream
                val writer = OutputStreamWriter(outputStream)
                writer.write(body.toString())
                writer.flush()
                writer.close()

                val responseCode = connection.responseCode
                Log.d("GoMailer", "📤 Event sent to backend. Status: $responseCode")

                connection.disconnect()
            } catch (e: Exception) {
                Log.e("GoMailer", "❌ Failed to send event: ${e.message}")
            }
        }.start()
    }
}

File: sdk/react-native/ios/GoMailer/GoMailerModule.swift
import Foundation
import React
import UserNotifications

@objc(GoMailerModule)
class GoMailerModule: RCTEventEmitter {
    
    private var apiKey: String?
    private var baseUrl: String = "https://api.go-mailer.com/v1"
    private var deviceToken: String?
    private var userEmail: String?
    
    override init() {
        super.init()
        print("🔵 GoMailerModule: Initialized")
    }
    
    @objc
    override static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    @objc
    override func supportedEvents() -> [String]! {
        return ["GoMailerDidRegisterForRemoteNotifications", "GoMailerDidFailToRegisterForRemoteNotifications"]
    }
    
    @objc(initialize:resolver:rejecter:)
    func initialize(_ config: [String: Any], resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: initialize called with config:", config)
        
        guard let apiKey = config["apiKey"] as? String else {
            reject("INVALID_CONFIG", "API key is required", nil)
            return
        }
        
        self.apiKey = apiKey
        if let baseUrl = config["baseUrl"] as? String {
            self.baseUrl = baseUrl
        }
        
        print("🔵 GoMailerModule: SDK initialized with API key:", apiKey.prefix(10) + "...")
        resolve(["success": true])
    }
    
    @objc(setUser:resolver:rejecter:)
    func setUser(_ user: [String: Any], resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: setUser called with user:", user)
        
        guard let email = user["email"] as? String else {
            reject("INVALID_USER", "Email is required", nil)
            return
        }
        
        self.userEmail = email
        print("🔵 GoMailerModule: User set to:", email)
        resolve(["success": true])
    }
    
    @objc(registerForPushNotifications:resolver:rejecter:)
    func registerForPushNotifications(_ params: [String: Any], resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: registerForPushNotifications called")
        
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) { granted, error in
            DispatchQueue.main.async {
                if granted {
                    print("🔵 GoMailerModule: Notification permission granted")
                    UIApplication.shared.registerForRemoteNotifications()
                    resolve(["success": true])
                } else {
                    print("🔴 GoMailerModule: Notification permission denied")
                    reject("PERMISSION_DENIED", "User denied notification permission", error)
                }
            }
        }
    }
    
    @objc(getDeviceToken:rejecter:)
    func getDeviceToken(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: getDeviceToken called")
        
        if let token = deviceToken {
            print("🔵 GoMailerModule: Returning cached device token")
            resolve(["token": token])
        } else {
            print("🔴 GoMailerModule: No device token available")
            reject("NO_TOKEN", "Device token not available", nil)
        }
    }
    
    @objc(trackEvent:resolver:rejecter:)
    func trackEvent(_ params: [String: Any], resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: trackEvent called with params:", params)
        
        guard let eventName = params["eventName"] as? String else {
            reject("INVALID_EVENT", "Event name is required", nil)
            return
        }
        
        let properties = params["properties"] as? [String: Any]
        
        // Send event to backend
        sendEventToServer(eventName: eventName, properties: properties)
        resolve(["success": true])
    }
    
    @objc(setUserAttributes:resolver:rejecter:)
    func setUserAttributes(_ attributes: [String: Any], resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: setUserAttributes called with attributes:", attributes)
        resolve(["success": true])
    }
    
    @objc(addUserTags:resolver:rejecter:)
    func addUserTags(_ tags: [String], resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: addUserTags called with tags:", tags)
        resolve(["success": true])
    }
    
    @objc(removeUserTags:resolver:rejecter:)
    func removeUserTags(_ tags: [String], resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: removeUserTags called with tags:", tags)
        resolve(["success": true])
    }
    
    @objc(setAnalyticsEnabled:resolver:rejecter:)
    func setAnalyticsEnabled(_ enabled: Bool, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: setAnalyticsEnabled called with enabled:", enabled)
        resolve(["success": true])
    }
    
    @objc(setLogLevel:resolver:rejecter:)
    func setLogLevel(_ level: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        print("🔵 GoMailerModule: setLogLevel called with level:", level)
        resolve(["success": true])
    }
    
    // MARK: - Device Token Handling
    @objc
    func didRegisterForRemoteNotifications(_ deviceToken: Data) {
        print("🔵 GoMailerModule: didRegisterForRemoteNotifications called")
        
        let tokenString = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
        self.deviceToken = tokenString
        print("🔵 GoMailerModule: Device token received:", tokenString)
        
        // Send device token to backend
        sendDeviceTokenToServer(tokenString)
        
        // Send event to React Native
        sendEvent(withName: "GoMailerDidRegisterForRemoteNotifications", body: ["deviceToken": tokenString])
    }
    
    @objc
    func didFailToRegisterForRemoteNotifications(_ error: Error) {
        print("🔴 GoMailerModule: didFailToRegisterForRemoteNotifications called with error:", error)
        
        // Send event to React Native
        sendEvent(withName: "GoMailerDidFailToRegisterForRemoteNotifications", body: ["error": error.localizedDescription])
    }
    
    private func sendDeviceTokenToServer(_ token: String) {
        guard let apiKey = apiKey, let email = userEmail else {
            print("🔴 GoMailerModule: Missing API key or user email")
            return
        }
        
        let url = URL(string: "\(baseUrl)/contacts")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        
        let bundleId = Bundle.main.bundleIdentifier ?? "com.gomailer.pushTestReactNativeExample"
        let payload: [String: Any] = [
            "email": email,
            "gm_mobi_push": [
                "deviceToken": token,
                "platform": "ios",
                "bundleId": bundleId,
                "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0"
            ]
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        } catch {
            print("🔴 GoMailerModule: Failed to serialize payload:", error)
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                print("🔴 GoMailerModule: Failed to send device token:", error)
            } else if let httpResponse = response as? HTTPURLResponse {
                print("🔵 GoMailerModule: Device token sent to server, status:", httpResponse.statusCode)
                if let data = data, let responseString = String(data: data, encoding: .utf8) {
                    print("🔵 GoMailerModule: Response:", responseString)
                }
            }
        }.resume()
    }
    
    private func sendEventToServer(eventName: String, properties: [String: Any]?) {
        guard let apiKey = apiKey, let email = userEmail else {
            print("🔴 GoMailerModule: Missing API key or user email")
            return
        }
        
        let url = URL(string: "\(baseUrl)/events/push")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        
        let bundleId = Bundle.main.bundleIdentifier ?? "com.gomailer.pushTestReactNativeExample"
        let payload: [String: Any] = [
            "email": email,
            "eventName": eventName,
            "properties": properties ?? [:],
            "platform": "ios",
            "bundleId": bundleId,
            "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0",
            "timestamp": ISO8601DateFormatter().string(from: Date())
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        } catch {
            print("🔴 GoMailerModule: Failed to serialize event payload:", error)
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                print("🔴 GoMailerModule: Failed to send event:", error)
            } else if let httpResponse = response as? HTTPURLResponse {
                print("🔵 GoMailerModule: Event sent to server, status:", httpResponse.statusCode)
                if let data = data, let responseString = String(data: data, encoding: .utf8) {
                    print("🔵 GoMailerModule: Event response:", responseString)
                }
            }
        }.resume()
    }
}
